Blog Post: How does mill build android apps

vaslabs opened the pull request:
An introductory blog post for Mill's android support. I tried to focus on celebrating mill's design in helping to add all this android support. I think I need a better title, please let me know your ideas and also feedback for the content.

Also added a page in the android docs for the native endless tunnel example (so as I can also reference it in the blog post).

lefou reviewed (commented):
Nice article. I think the diagrams are a real value, even to non-Mill users. I once had a rudimentary android build working with `sbuild` (my personal predecessor of Mill) around 2014, where Android build where simpler, but even then, it was hard to get these dependencies out from pure documentation. Reverse engineering is required to get the whole picture and your work here is really appreciated.

vaslabs commented on website/blog/modules/ROOT/pages/15-android-build-flow.adoc:
```diff
@@ -0,0 +1,181 @@
+- Instrumented tests (separate APK, own resources, manifests, dependencies)
+- Native code (NDK builds, CMake integration)
+
+We cover a lot of these architecture styles in the https://github.com/com-lihaoyi/mill/tree/main/example/androidlib[Android examples] and third party integration examples covering https://github.com/com-lihaoyi/mill/tree/main/example/thirdparty/android-compose-samples[Android Compose], https://github.com/com-lihaoyi/mill/tree/main/example/thirdparty/android-endless-tunnel[Android Native] and http ...
```
I know, I have a point above to fix them, I just placed them quickly while writing the article to avoid losing my train of thought, sorry :sweat_smile:

vaslabs commented:
Fixed the links and added a page in the android docs for the native endless tunnel example.

I shortened the diagram, but it still overflows

<details>

<img width="1903" height="980" alt="image" src="https://github.com/user-attachments/assets/c4ef456f-7748-4a03-9a29-357dd68668c2" />

</details>

if you have any ideas how to fix it, if it's an issue

vaslabs commented:
> Some comments on this:
> 
> 1. We should definitely have screenshots/demos of things that work: Emulator, IDE, etc. That makes it look a lot more "real" than just describing things in words.
> 2. Less generic Mill praise (e.g. "This was possible because Mill’s design, incremental, transparent, and extensible maps surprisingly well onto Android’s intricate build process."), more of a user-perspective of why they may be interested in this. Is it easier to set up? Does it build faster? Is it easier to extend? How can we demonstrate these attributes in a concrete way that people who don't care about Mill at all can understand and appreciate? This will help motivate the blog post and tell users why they should care about the technical accomplishment and implementation details discussed below
> 3. The `Try it Out` section needs complete, concrete, copy-paste-able commands rather than instructions like `Clone one of the example Android projects`. Download instructions should come with `curl` commands or clickable links. We need to hook people into trying this out before they get bored and leave, and asking them to go to another page of the docs and try to figure it out will result in a lot of people just giving up

Thanks. I've made the adjustments. In some place, I put comparison with Gradle and in other a generic "other build tools", I'm not sure which one you'd prefer, but when you get the chance, let me know of any more feedback

lihaoyi commented on website/blog/modules/ROOT/pages/15-android-build-flow.adoc:
```diff
@@ -0,0 +1,223 @@
+_{author}, {revdate}_
+
+
+Until recently, Gradle was the only realistic option for Android builds. Today, Mill - a fast, predictable build tool for Java, Scala, and Kotlin - can build Android applications end-to-end.
```
Let's leave out `fast, predictable` here. Whether or not it's true, it sounds too salesly. Let's just say "Mill provides an alternative to Gradle that is easier to use and learn", and hope that enough people have had trouble with Gradle that they empathize

lihaoyi commented on website/blog/modules/ROOT/pages/15-android-build-flow.adoc:
```diff
@@ -0,0 +1,223 @@
+Multi-module Pokedex sample app (https://github.com/NicosNicolaou16/Pokedex_Compose_Multi_Module[original], https://github.com/vaslabs/Pokedex_Compose_Multi_Module/tree/testing-mill[with mill])
+image:AndroidPokedexMultimoduleExample.png[A multi-module Android app built with Mill, showing a list of Pokémon and details for each pokemon.]
+
+Unlike Gradle, where build steps are buried inside plugins and DSL magic, Mill exposes each Android step as a task you can run, override, or extend.
```
Is there any angle we can discuss of Mill being easier to setup v.s. Gradle in terms of the number of manual steps? Or more performant, in terms of time-to-update-the-app? We should try to discuss some user-facing benefits in a way people can see for themselves why it's cool, rather than just "Mill is fast and predictable because we say so", so they are more engaged for the internal technical discussion below

lihaoyi commented on website/blog/modules/ROOT/pages/15-android-build-flow.adoc:
```diff
@@ -0,0 +1,223 @@
+
+Mill’s Android support is still young, but it already covers the full build pipeline: resource compilation, manifest merging, packaging, signing, running, and even testing on emulators.
+
+What makes this different from other build tools is control: every build step is a visible Mill task, easy to run on its own, inspect, or override. That means you can debug problems faster, adapt the pipeline to your project’s needs, and extend it without fighting opaque built-in or plugin logic.
```
The obvious question to me here is: is this not true of Gradle? Maybe it's not, but you have to show that in more detail for me to understand why Mill is cool

vaslabs commented:
thanks for the comments, I will come back to this tomorrow or the day after (just fyi)!

vaslabs commented:
I think I've addressed the comments, if you could do another review. There's time as I want to get the latest mill version in the [example](https://github.com/com-lihaoyi/mill/pull/5776/files#diff-2fdce5ff18ae58845b50ace9ecd6e12395cb3cc2bb7bddf168b29dc31a645581R209) so it's compatible with the build.mill I'm [referencing](https://github.com/com-lihaoyi/mill/pull/5776/files#diff-2fdce5ff18ae58845b50ace9ecd6e12395cb3cc2bb7bddf168b29dc31a645581R217) which I should also update to the most recent I can so it's easier for me to reason and test the content

vaslabs commented:
> @vaslabs now that 1.0.5 is out with your latest improvements to the android toolchain, shall we finalize this blog post and publish it?

Yes, I will get on with it today!

lefou commented:
IMHO an animated GIF is ok, but it should not change content multiple times a second. Give the reader some time to recognize the image, before changing it.

vaslabs commented:
> I'm getting an error trying to build the website on this PR locally
> 
> ```
> $ ./mill website.fastPages                                      
> [2641/2641, 1 failed] ============================== website.fastPages ============================== 1s
> 1 tasks failed
> [2151] example.kotlinlib.basic[2-module-deps].parsed java.lang.RuntimeException: No build file named build.sc/build.mill/build.mill.scala found in /Users/lihaoyi/Github/mill/out/example/kotlinlib/basic/2-module-deps/testRepoRoot.dest
>     scala.sys.package$.error(package.scala:27)
>     mill.testkit.ExampleParser$.$anonfun$2(ExampleParser.scala:14)
>     scala.Option.getOrElse(Option.scala:201)
>     mill.testkit.ExampleParser$.apply(ExampleParser.scala:15)
>     build_.example.package_.build_$example$package_$ExampleCrossModule$$_$parsed$$anonfun$1$$anonfun$1(package.mill:92)
>     mill.api.Task$Named.evaluate(Task.scala:338)
>     mill.api.Task$Named.evaluate$(Task.scala:323)
>     mill.api.Task$Computed.evaluate(Task.scala:349)
> ```

strange, I had it running let me try again

vaslabs commented:
only thing I can think of, is when I make some changes and don't track the directory, then mill is complaining with a similar error, could you double check with `git status` that it's clean?

vaslabs commented:
Ok, nice! it's ready then from my side!

lihaoyi commented:
@vaslabs yes there is version skew: on PRs we build everything against `HEAD`, but post-merge we build the blog against `main` and the docs against the latest stable (e.g. `1.0.5`). Perhaps you could remove the problematic link for the time being?
